Add `unique_shifts` method for `atomman.defects.FreeSurface`

lan496 opened the pull request:
`FreeSurface.shifts` returns candidates for terminations of free surfaces. However, Some of the terminations are identical under symmetry operations of `ucell`. This PR implements `unique_shifts` method which returns the nonequivalent terminations under symmetry operations. This function will be useful to avoid duplicated surface calculations.

Procedure for filtering nonequivalent shifts are as follows:
1. analyze a space group of `ucell` by `spglib`
2. create `Plane` objects to represent `shift` and apply symmetry operations in the space group to them
3. If planes are identical with some symmetry operation and lattice translation, choose one of them

I have added a unit test for (100) surfaces of anatase, which is a working example from Section 2.4 of [Sun and Ceder (2013)](https://www.sciencedirect.com/science/article/abs/pii/S003960281300160X?via%3Dihub). I have verified that `unique_shifts` function correctly returns the result with the article: there are two possible shifts and they are identical under a glide operation.

lan496 committed:
Add Box.cart method and testing for free surface

lan496 committed:
Add symmetry operation on plane

lan496 committed:
Add unique_shifts method for FreeSurface

FreeSurface.shifts returns candidates for terminations of free surfaces.
However, Some of the terminations are identical under symmetry operations of
ucell. This fix implements `unique_shifts` method which returns the
nonequivalent terminations under symmetry operations.

lmhale99 commented:
This is a good thing to add! I remember the Sun and Cedar paper mentioning this, so it’s good that someone figured out how to implement it.

I read through the code and have a few comments so far.  I’ll start running and testing later this week.

The Box.cart method is ambiguous and I think is being used differently than is defined. The description says conversions of coordinates from relative to Cartesian, but it is being used to convert crystal vectors to Cartesian.  They would be the same except Box origins are not explicitly (0,0,0).  The respective conversions are in atomman elsewhere (tools.miller.vector_crystal_to_cartesian and System.unscale), but could be added as methods of Box.  

I’ll work on the Box methods. Adding the vector conversion to Box is trivial. I need to think about renaming scale/unscale to something more informative like relative_to_cartesian or coord_relative_to_cartesian. It’s also curious how I seem to be doing effectively the same operation three different ways - NDArray.dot, np.inner with transpose, and sum over np.outer.

For Plane, my only comment is that maybe there should be another method “isclose” that does the plane comparisons but also allows atol and rtol values to be specified as users may want to change them. Then, either use Plane.isclose for the comparisons or make \_\_eq\_\_ call Plane.isclose using default tolerance values.

With FreeSurface, should there be a variable for the trial image range? int with default value of 1?

I should probably also start looking into adding annotations elsewhere.

lan496 committed:
Use proper conversion for lattice vectors

lan496 committed:
Add tolerance parameters to compare planes

lan496 committed:
Add parameter to enlarge trial images

lan496 commented:
Thanks for your comments. I have added some quick fixes. I greatly appreciate your reviews!

> I remember the Sun and Cedar paper mentioning this, so it’s good that someone figured out how to implement it.

I cannot find a relevant implementation to filter duplicated shifts. For example, [pymatgen’s surface module](https://github.com/materialsproject/pymatgen/blob/master/pymatgen/core/surface.py#L1096-L1169) enumerates all possible shifts and filter them after constructing structures by `StructureMatcher`.
So, there may be a better algorithm than that implemented in this PR.

> The respective conversions are in atomman elsewhere (tools.miller.vector_crystal_to_cartesian and System.unscale)
> I need to think about renaming scale/unscale to something more informative like relative_to_cartesian or coord_relative_to_cartesian.

Sorry, I did not notice this function. I have rewritten the code to properly use `vector_crystal_to_cartesian`.
For now, I tentatively leave `Box.cart` method. 

> For Plane, my only comment is that maybe there should be another method “isclose” that does the plane comparisons but also allows atol and rtol values to be specified as users may want to change them.

That’s good things to do! I have added `Plane.isclose` method and parameters to control the torelances from `FreeSurface.unique_shifts`.

> With FreeSurface, should there be a variable for the trial image range? int with default value of 1?

I have added `trial_image_range` parameter. BTW, I cannot come up with another way to find translationally equivalent planes. There may be a cleverer way not to use heuristics like trial_images.

lan496 commented:
> It’s also curious how I seem to be doing effectively the same operation three different ways - NDArray.dot, np.inner with transpose, and sum over np.outer.

This is a fun exercise! In my environment, `np.dot` is the fastest although all functions are so fast with small arrays. 

```python
import numpy as np
vects = np.random.rand(3, 3)
pos = np.random.rand(3)

def inner(vects, pos):
    return np.inner(pos, np.transpose(vects))

def dot(vects, pos):
    return np.dot(pos, vects)

def einsum(vects, pos):
    return np.einsum('ia,i->a', vects, pos)

%timeit inner(vects, pos)
> 2.41 µs ± 6.58 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit dot(vects, pos)
> 1.21 µs ± 2.25 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit einsum(vects, pos)
> 2.55 µs ± 11.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

lan496 committed:
Not to use am.rootdir

lan496 committed:
Fix parameters for atom_data dump in testing

lan496 committed:
Update checking for atomic datum

lmhale99 committed:
added potential parameter back in except for load

Leaving it in as additional tests may use potentials later.

lmhale99 committed:
Merge pull request #27 from lan496/fix-tests

Fix broken unit tests

lmhale99 committed:
Merge branch 'unique-shifts' of https://github.com/lan496/atomman into unique_shifts

lmhale99 committed:
plane_crystal_to_cartesian now handles arrays of planes, code cleanup

lmhale99 committed:
Conversion tools added as Box methods

lmhale99 committed:
scale and unscale methods labeled for future depreciation

lmhale99 committed:
System.unscale replaced with System.box.position_relative_to_cartesian

lmhale99 commented:
For the three dot product operations, I was also getting the array.dot() to be slightly faster than the others for small arrays.  Interestingly, it did seem like np.outer() with ellipsis summing and reshaping was faster for high dimensional arrays, but I doubt many people would be doing that.  With array.dot() also being the most concise, I started to clean up the related operations to use that.

With the Miller tools, I also saw that the plane_crystal_to_cartesian was only supporting a single set of plane indices, so I modified it to also support arrays of indices like the other operations.  

Box now has the miller conversions and updated forms of scale and unscale from System.  I went with calling the updated methods position_relative_to_cartesian and vice versa. The system methods now call the Box methods and are marked with future depreciation warnings.  

As for the trial images, the only other option I can think of is some normalization scheme to move all image shifts inside the box, but even then replicas may need to be tested if the shifts are exactly the lattice dimensions.

Tests and Jupyter docs still need to be added/updated for my changes...

lan496 commented on atomman/core/Box.py:
```diff
@@ -665,8 +680,111 @@ def inside(self, pos, inclusive=True):
+            raise ValueError('Invalid position dimensions')
+        
+        # Convert and return
+        return (value - self.origin).dot(np.linalg.inv(self.vects))
```
It will be a bit faster to precompute inverse of `self.vects` in setter of `vects` although calculating an inverse of such a tiny matrix does not take time.

lan496 reviewed (commented):
Looks good to me! I add a tiny comment on `Box.position_cartesian_to_relative`.

lmhale99 committed:
Only calculate inverse vects once

lmhale99 committed:
Add reciprocal_vects as class property

lmhale99 commented:
I started with making inverse_vects a private variable to avoid it being confused with reciprocal cells. Peeking at ase and pymatgen, it appears that reciprocal_vects is just the transpose of the inverse_vects, so I made it a property.  Not sure why ase uses pinv as inv seems to be equivalent but faster for all vects I tested. Hopefully I'm not missing anything important.

reciprocal_vects gets computed and stored the first time it is accessed, and is reset whenever vects is updated.  This minimizes the number of times it gets computed.  I saw roughly 7 microseconds of savings per call with the value saved, so it might be overkill.

lan496 reviewed (commented):
Thanks! `Box.reciprocal_vects` is very useful to me because I am sometimes not sure whether to use transpose or not...

As for ASE's `Cell`, `np.linalg.pinv` seems to be introduced in [this PR](https://gitlab.com/ase/ase/-/merge_requests/352).
`pinv` was introduced because they use [0, 0, 0] as an undefined lattice vector, which is not the case in atomman. So, I also think it is fine to use `np.linalg.inv` here.
